editor/extension.ts

```diff
@@ -270,8 +268,53 @@ namespace pxt.editor {
                 xyField.textContent = (xyField.textContent || "").replace(/^tilemap./, "tiles.");
             });
         }
+
+        if (pxt.semver.strcmp(pkgTargetVersion || "0.0.0", "1.12.34") < 0) {
```

Comment:
Just wanting to learn a bit about the larger `patchBlocks` function and why we have so many of these "if" statements. Are these needed because of our use of our fork of blockly, our extensive backwards compatibility, and translations getting broken because of blocks changes?

Reply:
Unrelated to us having a fork of blockly, just for backwards compat on our internal representation of the blocks code; this function runs whenever we load up a project to give us a chance to, well, patch the blocks code before it gets loaded into blockly / the editor. Each block of upgrade logic is an upgrade to the blocks xml to accommodate for a breaking change that applies depending on how long ago the project was created -- e.g. it's possible to load up a game made in pxt-arcade v0.2.1 into the live site, so any breaking changes we have made to block apis needs to be fixed (e.g. block names, new parameters, changes to the ways parameters are shown, etc).

In this case, https://github.com/microsoft/pxt/pull/9611 is a breaking change, because existing projects with those invalid translations do technically work in their current language, just not when shared / switching languages -- so these upgrade rules are here to make sure those projects continue to work in their current language, but also ideally in other languages when shared~

for ref, there's another place upgrade rules can be applied for typescript code, which is in pxtarget.json -- there's a compile field with a map of patches where the key is the range of projects from which the regex should be applied, and the value is an operation to apply (e.g. the "api" type for a regex replace).

Reply:
Cool! Thank you for the in-depth description!